### DISCUSSION TITLE
调用 `cusdoc.cls` 写手册过程中的一些问题和需求

### DISCUSSION THREAD

**Discussion by xkwxdyy** [Sophanatprime/cus#40] [General]
#34 后 `cus` 开发了 `cusdoc` 文类，然后本 discussion 主要讨论使用过程中遇到的一些问题和需求。

**Comment by xkwxdyy** [comment]
## 如何排出“模板构成”的效果

用户手册中一般会包括向用户介绍模板的文件构成，比如 Markdown 的：
```md
asset
|--logo.png

foo.cls

demo.tex
```
类似于这种，但是不知道如何用 `LaTeX` 排版出一种介绍构成的效果，感觉有点像 `cus-cn` 里介绍目录定制的时候的那个？但不知道能否包装优化一下，生成一个新的环境或者命令来排版这种需求。

**Reply by Sophanatprime** [reply]
已经有宏包，比如 `dirtree`，`forest` 等，可以排版出这样的效果。

**Reply by xkwxdyy** [reply]
好的，感谢！

**Comment by xkwxdyy** [comment]
## 关于文件夹的输出
比如 `assets` 目录，`body` 目录。

这个好像也不算 file 吧？感觉用 `\file{assets}` 怪怪的，

**Reply by Sophanatprime** [reply]
`\file` 只是 `\nolinkurl` 的别名，如果觉得不合语意，可以用它定义新的合适的命令。

**Reply by xkwxdyy** [reply]
好的，本来最好是 `\path`，但是这个被 `tikz` 用了。

**Comment by xkwxdyy** [comment]
## 排版键值对

目前来说排版 `<key>=<val>` 的效果不太方便，可能要 `\meta{key} = \meta{val}`，貌似目前没有一个排版键值对的功能。

可以像 `fdudoc.cls` 一样[搞一个类似于 `\kvopt` 的命令](https://github.com/stone-zeng/fduthesis/blob/d84e0dd7df657d47943e98c98ff843cc1336bc5f/source/fduthesis-doc.dtx#L1901-L1907)

**Reply by Sophanatprime** [reply]
在 `syntax` 环境可以直接写 `key=<{var}>`，没懂哪里不方便。而且 `\meta` 是表示值的。

**Reply by xkwxdyy** [reply]
嗯嗯，我是指在 `syntax` 外使用，因为一般需要在外面进行一些解释。

**Reply by xkwxdyy** [reply]
我看您一般就是直接 `\verb|key=val|` 这样写的

**Reply by Sophanatprime** [reply]
这个应该自己封装一下  `\opt` `\key` `\meta` 就行。

**Reply by xkwxdyy** [reply]
好的，我只是觉得这个可能可以作为 `cusdoc.cls` 的其中一个封装命令

**Comment by xkwxdyy** [comment]
## 如何手册里记录版本控制

我看到 `cus-cn.tex` 里有 `\PrintChanges` 命令，但是并不太清楚怎么记录 changes， `cus-cn.tex` 里面也没有相关的例子。

**Reply by xkwxdyy** [reply]
在前面加了 `\changes{v0.2}{2024/04/17}{完成手册的编写。}`，然后最后
```latex
\PrintChanges 
```
可以显示“版本历史”的标题，但是没有内容，是我的使用方式不当吗？

**Reply by Sophanatprime** [reply]
`\PrintChanges` 和 `\PrintUsages` 都需要执行 `makeindex`，编译两次才行。看看 log 里有没有
```log
runsystem(makeindex -s gglo.ist ...)
```

**Reply by Sophanatprime** [reply]
如果没有自动执行 `makeindex`，看看是不是设置了 `\setupindex[docchange]{auto=false}`。

**Reply by xkwxdyy** [reply]
```latex
\newindextype[auto=true,filename=\jobname.idx,heading*={\section}]{\empty}
\setupindex[\empty,docchange]{auto=false}
```
对的， `cus-cn.tex` 里有这样一段

**Reply by xkwxdyy** [reply]
log 里有一个 `runsystem(makeindex -s gind.ist -o Now-Learn-English-doc.docusage.ind`

然后我在命令行使用了这行代码，
```bash
☁  doc [main] ⚡  makeindex -s gind.ist -o Now-Learn-English-doc.docusage.ind
Scanning style file /usr/local/texlive/2024/texmf-dist/makeindex/latex/gind.ist..........
** Input style error (file = /usr/local/texlive/2024/texmf-dist/makeindex/latex/gind.ist, line = 75):
   -- Unknown specifier lethead_prefix.
** Input style error (file = /usr/local/texlive/2024/texmf-dist/makeindex/latex/gind.ist, line = 76):
   -- Unknown specifier lethead_suffix.
** Input style error (file = /usr/local/texlive/2024/texmf-dist/makeindex/latex/gind.ist, line = 77):
   -- Unknown specifier lethead_flag.
...done (13 attributes redefined, 3 ignored).
This is makeindex, version 2.17 [TeX Live 2024] (kpathsea + Thai support).
Scanning input file stdin...
```
然后就一直卡在这不动，好像有几分钟了

**Reply by xkwxdyy** [reply]
我把上面的 `makeindex` 终止了，然后把 `\setupindex[\empty,docchange]{auto=false}` 改成 `\setupindex[\empty,docchange]{auto=true}` 之后 `xelatex` 后发现版本历史编译出来了。
![image](https://github.com/Sophanatprime/cus/assets/87134063/147b3a89-22d4-4449-8946-8e8de233c038)

**Reply by Sophanatprime** [reply]
`makeindex` 没有指定输入的 `idx` 文件。
应该是这样
```
makeindex -s gind.ist -o Now-Learn-English-doc.docusage.ind Now-Learn-English-doc.docusage.idx
```

一般让它自动执行就好了。不需要开启 `--shell-escape` 就能执行。

**Reply by xkwxdyy** [reply]
感谢！出来了。原来是要用手动用 `\index` 先录入，第一次用，不太熟悉，抱歉。我以为和代码索引一样是自动出来的。想想也是不录入的话输出啥呢

**Reply by Sophanatprime** [reply]
版本历史是需要手动添加 `\changes` 的，使用 `function` `keyval` `\cs` `\tn` `\cmd` `\key` 的命令和键都会自动加到代码索引中。
只需要设置 `\setupindex[docusage,docchange]{auto=true}`，编译两次就能得到。

自己定义的索引，需要使用 `\index` 添加到对应的索引文件中才行。

**Reply by xkwxdyy** [reply]
`\changes{v0.2}{2024/04/17}{完成手册的编写。}` 我这样输入的，然后输出是
![image](https://github.com/Sophanatprime/cus/assets/87134063/e6dced19-a4ec-4516-becd-b0d3e5eef7da)

在不同地方写 `\changes` 是会改变 `General` 吗？一般在什么地方写 `\changes` 呢？

**Reply by Sophanatprime** [reply]
只会在 `.dtx` 的 `macro` 环境里才会修改它，用于表示哪个命令被修改了，效果可以参考 source2e.pdf。因为 `doc` 库不支持 `.dtx`，所以总是不会变的。

**Reply by Sophanatprime** [reply]
一般在命令或环境的说明处写，也可以统一在一个地方写。

另外，`cus` 在 0.1 版本之后才会添加版本历史。

**Reply by xkwxdyy** [reply]
了解了，我想去掉`General:`，我在 `source2e.tex` 查到这样的代码：
```latex
\def\changes@#1#2#3{%
  \let\protect\@unexpandable@protect
  \edef\@tempa{\noexpand\glossary{\@changes@iso@date#2///\space
                                 \currentfile\space#1\levelchar
                                 \ifx\saved@macroname\@empty
                                   \space
                                   \actualchar
                                   \generalname
                                 \else
                                   \expandafter\@gobble
                                   \saved@macroname
                                   \actualchar
                                   \string\verb\quotechar*%
                                   \verbatimchar\saved@macroname
                                   \verbatimchar
                                 \fi
                                 :\levelchar #3}}%
  \@tempa\endgroup\@esphack}
```

然后我用 `\def\generalname{}` 发现可以去掉 `General`，但是后面的 `:` 没有去掉，然后我把上面的这段代码放在 `cus-cn.tex` 导言区 ` :\levelchar #3}}%` 改为 ` \levelchar #3}}%`，然后去掉了 `\generalname`，但是编译报错
```latex
Undefined control sequence.
\changes@ ...expand \glossary {\@changes@iso@date
```
`source2e` 我以为都在内核里，怎么也会这样呢，是我操作有误吗？我前面加了 `\makeatletter`的，有什么更好的办法去掉 `General:` 吗

**Reply by Sophanatprime** [reply]
`source2e.pdf` 使用 `l3doc` 排版，`\changes` 并不在内核里。

https://github.com/Sophanatprime/cus/commit/e497371ac6ce31d2117f9bd8b5f377c668b1a1f2 增加了 `\removegeneralname`，使用后，会移除它后面的 `\changes` 所加的 “General”。

**Reply by xkwxdyy** [reply]
感谢！

**Comment by xkwxdyy** [comment]
## 关于 `\DocumentMetadata`

请问这个有什么用呢？我看到手册里好像有一句说加这个好点，但我第一次见这个命令，而且还放在 `\documentclass` 开头

**Reply by Sophanatprime** [reply]
用于启用 “new LaTeX PDF management”，主要是和 PDF 文件格式有关的特性，可参考 pdfmanagement-testphase.pdf。

设置透明度的 `l3opacity` 需要启用它，XeLaTeX 需要使用它才能在正文中使用透明度（只在 tikz 图片中则不需要）。

`cus` 的 `pdfmana` 模块和 `pdf` 库需要启用它才能生效。目前提供设置透明度和导入 PDF 保留超链接的功能。

**Reply by Sophanatprime** [reply]
它只能放在 `\documentclass` 命令之前。

**Reply by xkwxdyy** [reply]
了解了，感谢。

**Comment by xkwxdyy** [comment]
## `function` 的效果优化

![image](https://github.com/Sophanatprime/cus/assets/87134063/4b0b67d2-cfa3-4183-9a8a-fc18e1a39158)

这张图可以看到目前的效果的不足，解释性文字和对应的内容关联性不强，特别是没有 `syntax` 的时候就更明显了，这个图片里的解释性文字紧跟在一个例子后面，我刚开始阅读的时候以为这是例子的进一步解释，结果读着读着发现不对，看旁边才知道已经到下面的部分了。

另外，我个人觉得 `function` 的文字部分可能视觉上也希望优化一下
![image](https://github.com/Sophanatprime/cus/assets/87134063/c1bb2b30-db3e-4862-8d1b-e9e95a24bbbf)

一般来说，第一段不缩进的用于简洁解释，而第二段缩进的为详细解释或者是补充内容。这两段的性质有一点区别，可能用某种方式区分，比如给第一段加个浅灰色背景之类的。

**Reply by Sophanatprime** [reply]
第一个确实有需要，而且实现也不难。

第二个，缩进与否就已经做了区分，是否需要进一步的区分还有待考虑。

**Reply by xkwxdyy** [reply]
嗯嗯，第二个可能只是我不太习惯这样子的排版效果，最终还是看您的处理。

**Comment by xkwxdyy** [comment]
## `keyval` 环境不能跨行

MWE：
```latex
\DocumentMetadata{}
\documentclass{cusdoc}

\newindextype[auto=true,filename=\jobname.idx,heading*={\section}]{\empty}
\setupindex[\empty,docchange]{auto=true}

\def\breakablethinspace{\hskip 0.16667em\relax}
\DeclareDocumentCommand\kv{mm}
  % {\texttt{#1\breakablethinspace=\breakablethinspace#2}}
  {\key{#1}\breakablethinspace=\breakablethinspace\opt{#2}}


\begin{document}
\begin{keyval}[path=env, added = 2024-04-29]{show-list,show-list+}
  \begin{syntax}
    show-list = {<env list>}
    show-list+ = {<env list>}
  \end{syntax}
  控制定理类环境的显隐，在 \meta{env list} 里的环境才会显示。\key{show-list} 是重置，而 \key{show-list+} 是在已有的基础上增加。

  目前支持的环境有：
  \begin{itemize}
    \item \env{definition}
    \item \env{theorem}
    \item \env{corollary}
    \item \env{aims}
    \item \env{explore}
    \item \env{example}
    \item \env{variation}
    \item \env{tips}
    \item \env{expansion}
    \item \env{postscript}
    \item \env{note}
    \item \env{keypoint}
    \item \env{exercise}
  \end{itemize}
  
  \kv{version}{teaching} 和 \kv{version}{studying} 两个版本预设了两种不同的 \key{show-list}:
  \begin{itemize}
    \item \kv{version}{teaching}：
    \begin{itemize}
      \item \env{definition}
      \item \env{theorem}
      \item \env{corollary}
      \item \env{aims}
      \item \env{explore}
      \item \env{example}
      \item \env{variation}
      \item \env{tips}
      \item \env{expansion}
      \item \env{postscript}
      \item \env{note}
    \end{itemize}
    \item \kv{version}{studying}：
    \begin{itemize}
      \item \env{example}
      \item \env{variation}
      \item \env{expansion}
      \item \env{keypoint}
      \item \env{exercise}
      \item \env{example}
      \item \env{variation}
      \item \env{expansion}
      \item \env{keypoint}
      \item \env{exercise}
      \item \env{example}
      \item \env{variation}
      \item \env{expansion}
      \item \env{keypoint}
      \item \env{exercise}
      \item \env{example}
      \item \env{variation}
      \item \env{expansion}
      \item \env{keypoint}
      \item \env{exercise}
    \end{itemize}
  \end{itemize}

  如果是默认的处理应该在 \file{teaching-plan.cls} 里修改预设（但需要告知我，或者让我来改，否则之后的更新会替换掉你之前的修改），如果是临时地修改显隐就在 \file{.tex} 里使用 \keyreflist[env]{show-list,show-list+} 来修改即可。
\end{keyval}

\end{document}
```

![image](https://github.com/Sophanatprime/cus/assets/87134063/c1a15dda-b372-4455-bb58-b0d3e067bdbb)

发现无法断页，可能是最新版本加了颜色导致的？

**Reply by Sophanatprime** [reply]
原本就不能断页。l3doc，ctxdoc 都是如此。

**Reply by xkwxdyy** [reply]
这个 `cus` 能否优化呢？不能的话我之后用就注意一下，只把主要内容放里面，其它放 `function` 外面

**Reply by Sophanatprime** [reply]
虽然可以做到，但比较复杂，还是保持 `function` 简短为好。

**Reply by xkwxdyy** [reply]
好的，谢谢

**Comment by xkwxdyy** [comment]
## `\keyreflist` 的换行问题

```latex
\keyref[solution]{vsep} 主要是用于临时调整，和题型无关，而 \keyreflist[solution]{extra-skip-choice,extra-skip-fillin,extra-skip-problem} 用于设置不同题型下的默认额外间距。
```
![image](https://github.com/Sophanatprime/cus/assets/87134063/973d5d45-e54e-463f-9b7e-68a4630f0877)

发现 `extra-skip-choice` 没有断行，我看 `cus-cn.tex` 里没有对 `\keyreflist` 的断行处理，请问有没有手工断行的办法呢

**Reply by xkwxdyy** [reply]
改成
```latex
\begin{itemize}
  \item \keyref[solution]{vsep} 主要是用于临时调整，一般和题型无关
  \item \keyreflist[solution]{extra-skip-choice,extra-skip-fillin,extra-skip-problem} 用于设置不同题型下的默认额外间距。
\end{itemize}
```
出现了另一种不能断行的情形：
![image](https://github.com/Sophanatprime/cus/assets/87134063/e5c6afb4-5f73-4a9c-9303-dae8a6557010)

**Reply by Sophanatprime** [reply]
手动添加 `\-`，或设置 `breat-at-any=true`。

**Reply by xkwxdyy** [reply]
好的谢谢。我查手册的时候也想用 `breat-at-any=true` 的，但是这个好像是 `cmd` 的可选参数，我看了 `keyref` 的定义，貌似除了 path 就没有其它可选接口了，所以一开始就没用这个。

**Reply by Sophanatprime** [reply]
`\keyref` 是不能直接设置键值选项的，要用 `\cussetup` 或者用 `\key`。

**Reply by xkwxdyy** [reply]
好的，谢谢

**Comment by xkwxdyy** [comment]
## 引用 `xample` 环境时 `\label` 也出现在代码中
```latex
\DocumentMetadata{}
\documentclass{cusdoc}

\begin{document}

引用测试~\ref{example:test}

\begin{xample}\label{example:test}
测试 \emph{测试}
\stopxamplecode 
\xampleprint 
\end{xample}

\end{document}
```
![image](https://github.com/Sophanatprime/cus/assets/87134063/6d1e7c07-282d-4384-8950-daa7a3c4f512)
虽然引用正确，但是 `\label{example:test}` 也出现了，这个可以修复吗？

**Reply by Sophanatprime** [reply]
放在 `\stopxamplecode` 的下一行。

**Reply by xkwxdyy** [reply]
感谢
